### PR TITLE
PERF-3502 PPaaS - Create Web component Unit Tests for Controller

### PR DESCRIPTION
**Pull request by tkmcmaster**
## Summary

Adds comprehensive Vitest-based unit tests for 22 controller React components, migrates coverage reporting to a merged Mocha + Vitest pipeline, and delivers mobile layout and UX improvements to the `TestResults` and `TestResultsCompare` components.

## Motivation

The controller's component layer had no unit test coverage beyond integration-level Mocha tests in `controller/test/`. PERF-3502 called for component-level tests to catch regressions earlier and support ongoing UI development. Two sets of UI issues were also addressed opportunistically during this work: the results tables were unreadable on mobile (vertical letter-stacking in headers), and the `TestResultsCompare` column-visibility dropdowns operated independently instead of staying in sync.

### Vitest infrastructure
- `controller/vitest.config.ts` — jsdom environment, globals, path aliases, custom coverage provider
- `controller/tsconfig.vitest.json` — extends main tsconfig, adds `vitest/globals` types
- `controller/vitest.setup.ts` — imports `@testing-library/jest-dom` matchers
- `controller/coverage-provider.ts` — wraps `@vitest/coverage-v8`; copies raw V8 JSON to `coverage/tmp/` before cleanup so `c8` can produce a merged final report across both runners
- `controller/components/tsconfig.json` — VS Code IntelliSense for spec files
- `controller/package.json` — new `testonly:mocha` / `testonly:components` scripts; `testonly` now runs both runners then `c8 report`; added devDeps: `vitest`, `@vitest/coverage-v8`, `@testing-library/jest-dom`, `@testing-library/user-event`
- `package.json` — removed `jsdom-global` from root (no longer needed)

### New component unit tests (30 spec files)
Tests follow Vitest + React Testing Library conventions with `jsdom`. All specs live co-located with their component under `controller/components/`.

| Component | File |
|---|---|
| Alert | `alert.spec.tsx` |
| Calendar | `calendar.spec.tsx` |
| CheckboxButton | `checkbox-button.spec.tsx` |
| Div | `div.spec.tsx` |
| DropFile | `drop-file.spec.tsx` |
| EnvironmentVariableItem | `env-var-item.spec.tsx` |
| EnvironmentVariablesList | `env-vars-list.spec.tsx` |
| FilesList | `files-list.spec.tsx` |
| Headers | `headers.spec.tsx` |
| Layout | `layout.spec.tsx` |
| LinkButton | `link-button.spec.tsx` |
| Modal | `modal.spec.tsx` |
| PewPewVersions (component + initialProps) | `pewpew-versions.spec.tsx`, `initialProps.spec.ts` |
| StartTestForm | `start-test-form.spec.tsx` |
| Table | `table.spec.tsx` |
| TestInfo | `test-info.spec.tsx` |
| TestQueues (component + initialProps) | `test-queues.spec.tsx`, `initialProps.spec.ts` |
| TestResults (charts, comparison, model, utils, component, quad panel charts) | 6 spec files |
| TestResultsCompare | `testresultscompare.spec.tsx` |
| TestsList | `tests-list.spec.tsx` |
| Toaster | `toaster.spec.tsx` |
| YamlViewer | `yaml-viewer.spec.tsx` |

### Removed legacy test files
- `controller/test/testresults.spec.tsx` — superseded by co-located spec
- `controller/test/testresultscompare.spec.tsx` — superseded by co-located spec

### Mobile layout fixes — TestResults and TestResultsCompare
- `QUADGRID` / `COMPARISONCHARTSGRID`: collapse from 2-column to 1-column grid at ≤768 px
- `TH`: `white-space: nowrap` — headers stay on one line; `overflow-x: auto` on the container handles horizontal scroll
- `DATATD`: `white-space: nowrap` + `overflow: hidden` + `text-overflow: ellipsis` — cells truncate cleanly instead of stacking vertically
- `FLEXROW` (styled.tsx): added `flex-wrap: wrap` so per-endpoint RTT/status charts stack on narrow screens
- `CANVASBOX`: responsive width via media query on mobile
- `DATATABLE` (compare): removed `table-layout: fixed` so columns size naturally to content

### TestResultsCompare UX improvements
- **Synchronized column dropdowns**: lifted `visibleColumns` state from `FinalResultsTable` to the parent `TestResultsCompare`; both Final Results tables now share one state object, so toggling a column in either dropdown immediately reflects in both
- **Metric row selector on Endpoint Comparison tab**: added `INITIAL_VISIBLE_METRICS` constant and a "Show Metrics" dropdown that controls which metric rows (Calls, Avg, Min, Max, StdDev, P50, P90, P95, P99) are visible across all endpoint comparison tables; defaults to Calls, Avg, P50, P95, P99

### Medium Priority Review Areas
- **`controller/coverage-provider.ts`**: Custom Vitest coverage provider copies raw V8 JSON before cleanup to enable merged reporting. The file path for the temp copy (`coverage/tmp/`) is hardcoded — verify this path is gitignored and doesn't conflict with CI artifact collection.
- **`controller/components/TestResultsCompare/index.tsx` (state lifting)**: Lifting `visibleColumns` out of `FinalResultsTable` changes the component contract. If `FinalResultsTable` is rendered elsewhere with independent column state expected, that assumption is now broken — verify no other callers exist.

### Testing Gaps
- The new Vitest specs cover render and interaction behaviour but do not exercise the WASM histogram path (`hdr-histogram-wasm`) — those calls are mocked. Real histogram behaviour is still only tested by the existing Mocha integration suite.
- `StartTestForm` spec is the most complex new test (278 lines); edge cases around file upload and environment variable mutation are worth a close read.

## Review Hotspots

1. **`controller/coverage-provider.ts`**
   - Why: Novel infrastructure — custom Vitest coverage provider is uncommon and required iteration to get the V8 JSON copy timing right before the runner cleans up.
   - Focus on: Verify the `beforeAll`/`afterAll` hook ordering ensures the raw JSON is copied before `@vitest/coverage-v8` deletes its temp files.

2. **`controller/components/TestResultsCompare/index.tsx` — state lifting + metric dropdown**
   - Why: Two separate concerns changed in one commit (synchronization + new feature); the component went through multiple iterations in this session.
   - Focus on: The `useEffect` for closing the metrics dropdown on outside-click uses a CSS class selector (`metric-select-container`) — confirm it doesn't conflict with the existing `column-select-container` handler now that both effects run in the same component.

3. **`controller/components/StartTestForm/start-test-form.spec.tsx`**
   - Why: Largest single spec file (278 lines); tests a form with async file upload, environment variable mutation, and complex prop-driven state.
   - Focus on: Verify mocks are scoped correctly and don't bleed between tests.

## Testing

- All 30 new Vitest spec files pass (`npm run testonly:components`)
- Existing Mocha integration suite unaffected (`npm run testonly:mocha`)
- `npm run linterror` passes with zero warnings
- Mobile layout verified visually against the screenshot provided during development
- Synchronized column dropdowns and metric row selector verified via Storybook stories

## Checklist
- [x] Code follows project style guidelines (`npm run linterror` passes)
- [x] Tests added for all new component behaviour
- [x] Legacy duplicate test files removed
- [x] No breaking changes to public component APIs
- [x] Reviewed own code for obvious issues

### CODE REVIEW

**Comment by tkmcmaster** [comment]
Merged controller coverage:
image
image

File: controller/vitest.config.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    include: ["components/**/*.spec.tsx", "components/**/*.spec.ts"],
+    setupFiles: ["./vitest.setup.ts"],
+    maxWorkers: 4,
+    coverage: {
+      provider: "custom",
+      customProviderModule: "./coverage-provider",
+      reporter: ["lcov", "text"],
+      reportsDirectory: "coverage/vitest",
+      include: ["components/**/*"],
+      exclude: ["**/story.tsx", "**/*.backup", "**/*.json"]
```
[the remark this answers is not in the document]
Comment (tkmcmaster) [reply controller/vitest.config.ts]: Good catch — added `**/*.spec.ts` and `**/*.spec.tsx` to `coverage.exclude` so spec files are excluded from the coverage report.

File: controller/package.json [hunk 2]
```diff
@@ -15,15 +15,17 @@
     "fix:install": "rimraf \"node_modules/@types/react-native/\"",
     "build:test": "npm run fix:install && tsc -p tsconfig.test.json --noEmit",
     "build:wdio": "tsc -p ./acceptance/wdio/tsconfig.json --noEmit",
-    "test": "npm run build:test && npm run build:wdio && c8 mocha --project tsconfig.test.json ./test",
-    "testonly": "c8 mocha --project tsconfig.test.json ./test",
+    "test": "npm run build:test && npm run build:wdio && npm run testonly",
+    "testonly": "npm run testonly:mocha && npm run testonly:components && c8 report",
+    "testonly:mocha": "c8 mocha --project tsconfig.test.json ./test",
+    "testonly:components": "vitest run --coverage --passWithNoTests",
     "acceptance": "npm run acceptance:api && npm run acceptance:wdio",
```
[the remark this answers is not in the document]
Comment (tkmcmaster) [reply controller/package.json:22]: Removed `--passWithNoTests` — component tests are expected to exist, so missing spec files should fail fast.

File: controller/components/DropFile/drop-file.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  describe("Drag interaction", () => {
+    it("should call onDropFile when files are dropped", () => {
+      const onDropFile = vi.fn();
+      render(<DropFile onDropFile={onDropFile} />);
+      const file = new File(["content"], "test.yaml", { type: "text/plain" });
+      const dropzone = screen.getByTestId("dropzone");
+      fireEvent.drop(dropzone, { dataTransfer: { files: [file] } });
+      // The Dropzone mock renders children — onDrop is passed via Dropzone component
+      expect(screen.getByTestId("dropzone")).toBeInTheDocument();
+    });
```
[the remark this answers is not in the document]
Comment (tkmcmaster) [reply controller/components/DropFile/drop-file.spec.tsx:68]: Fixed — rewrote the mock to capture `onDrop` in a module-level variable, then call it directly in the test and assert `onDropFile` was called with the dropped files.

File: eslint.config.mjs [hunk 4]
```diff
@@ -28,6 +28,10 @@ export default [{
     "controller/storybook-static",
     "controller/test/**.js",
     "controller/next-env.d.ts",
+    "controller/coverage-provider.ts",
+    "controller/vitest.config.ts",
+    "controller/vitest.setup.ts",
+    "controller/next-env.d.ts",
     "eslint.config.mjs"
```
[the remark this answers is not in the document]
Comment (tkmcmaster) [reply eslint.config.mjs:34]: Removed the duplicate entry.

File: controller/vitest.setup.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+import "@testing-library/jest-dom";
```
[the remark this answers is not in the document]
Comment (tkmcmaster) [reply controller/vitest.setup.ts]: Switched to `@testing-library/jest-dom/vitest` so matchers register with Vitest's `expect`.

File: controller/tsconfig.vitest.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
+{
+  "extends": "./tsconfig.json",
+  "compilerOptions": {
+    "types": ["vitest/globals", "@testing-library/jest-dom"]
+  },
```
[the remark this answers is not in the document]
Comment (tkmcmaster) [reply controller/tsconfig.vitest.json:5]: Updated to `@testing-library/jest-dom/vitest` to match the setup file.

File: controller/tsconfig.vitest.json [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+  "include": [
+    "vitest.setup.ts",
+    "vitest.config.ts",
+    "coverage-provider.ts",
+    "components/**/*.spec.ts",
+    "components/**/*.spec.tsx",
+    "components/**/*.tsx",
```
[the remark this answers is not in the document]
Comment (tkmcmaster) [reply controller/tsconfig.vitest.json:19]: Added an explicit `exclude` with only `node_modules` and the non-Vitest test directories, so spec files and `vitest.setup.ts` aren't excluded from the Vitest TS project.

File: controller/vitest.config.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
+import path from "path";
+import { defineConfig } from "vitest/config";
+
+export default defineConfig({
+  resolve: {
+    alias: {
+      "@fs/hdr-histogram-wasm": path.resolve(__dirname, "__mocks__/hdr-histogram-wasm.ts")
+    }
```
[the remark this answers is not in the document]
Comment (tkmcmaster) [reply controller/vitest.config.ts]: Switched to `fileURLToPath(import.meta.url)` to define `__dirname` — the esbuild shim happened to work locally, but this is the correct ESM pattern.

**Review by bryan-e-lopez — APPROVED** [review]
